From 2b8b577aecaffdd95c2f4b61c0b9dadfe29a2a65 Mon Sep 17 00:00:00 2001
From: Emilia Paz <emiliapaz@chromium.org>
Date: Mon, 16 Oct 2023 21:37:24 +0000
Subject: [PATCH] [M118-Extensions] Don't update access if site permissions aren't allowed

Page access options in the extension's context menu should only be
enabled when the extension site permissions can be changed. However, sometimes the command still gets invoked (crbug.com/1468151). Thus, we
exit early to prevent any crashes.

(cherry picked from commit bdf53322264cca476fd213da95552f011c13df84)

Bug: 1468151
Change-Id: I6549e9e95008873a2175a46d583c118ba92913f8
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/4874994
Commit-Queue: Emilia Paz <emiliapaz@chromium.org>
Reviewed-by: David Bertoni <dbertoni@chromium.org>
Cr-Original-Commit-Position: refs/heads/main@{#1198540}
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/4943086
Auto-Submit: Emilia Paz <emiliapaz@chromium.org>
Reviewed-by: Devlin Cronin <rdevlin.cronin@chromium.org>
Cr-Commit-Position: refs/branch-heads/5993@{#1309}
Cr-Branched-From: 511350718e646be62331ae9d7213d10ec320d514-refs/heads/main@{#1192594}
---

--- a/chrome/browser/extensions/extension_context_menu_model.cc
+++ b/chrome/browser/extensions/extension_context_menu_model.cc
@@ -429,7 +429,7 @@
     case PAGE_ACCESS_RUN_ON_CLICK:
     case PAGE_ACCESS_RUN_ON_SITE:
     case PAGE_ACCESS_RUN_ON_ALL_SITES: {
-      // If the web contents have navigated to a different origin, do nothing.
+      // Do nothing if the web contents have navigated to a different origin.
       auto* web_contents = GetActiveWebContents();
       if (!web_contents ||
           !origin_.IsSameOriginWith(web_contents->GetLastCommittedURL())) {
@@ -437,6 +437,16 @@
       }
 
       LogPageAccessAction(command_id);
+
+      // Do nothing if the extension cannot have its site permissions updated.
+      // Page access option should only be enabled when the extension site
+      // permissions can be changed. However, sometimes the command still gets
+      // invoked (crbug.com/1468151). Thus, we exit early to prevent any
+      // crashes.
+      if (!PermissionsManager::Get(profile_)->CanAffectExtension(*extension)) {
+        return;
+      }
+
       SitePermissionsHelper permissions(profile_);
       permissions.UpdateSiteAccess(*extension, web_contents,
                                    CommandIdToSiteAccess(command_id));
